src/ExactDiagonalization/basis_bfs.jl

```diff
@@ -0,0 +1,365 @@
+            sizehint!(T[], col_hint),
+            sizehint!(Dict{A,T}(), col_hint),
+            sizehint!(Set{A}(), col_hint),
+        )
+    end
+end
+
+function init_accumulator(::Type{<:MatrixBuilder{<:Any,T,I}}) where {I,T}
+    return MatrixBuilderAccumulator{T,I}()
+end
+
+function (builder::MatrixBuilder{<:Any,T})(operator, columns, seen) where {T}
+    empty!(builder.is)
+    empty!(builder.js)
+    empty!(builder.vs)
+    empty!(builder.frontier)
+    for (col_key, col_index) in columns
+        empty!(builder.column_buffer)
+        diag = diagonal_element(operator, col_key)
+        if !iszero(diag)
+            builder.column_buffer[col_key] = diag
+        end
+        for (row_key, value) in offdiagonals(operator, col_key)
+            if !iszero(value)
+                new_value = get(builder.column_buffer, row_key, zero(T)) + value
+                builder.column_buffer[row_key] = new_value
+                row_key ∉ seen && push!(builder.frontier, row_key)
+            end
+        end
+
+        old_len = length(builder.is)
+        new_len = length(builder.is) + length(builder.column_buffer)
+        resize!(builder.is, new_len)
+        resize!(builder.js, new_len)
+        resize!(builder.vs, new_len)
+        @inbounds for (i, (row_key, value)) in enumerate(builder.column_buffer)
+            builder.is[old_len + i] = row_key
+            builder.js[old_len + i] = col_index
+            builder.vs[old_len + i] = value
+        end
+    end
+end
+
+struct MatrixBuilderAccumulator{T,I}
+    is::Vector{I}
+    js::Vector{I}
+    vs::Vector{T}
+
+    MatrixBuilderAccumulator{T,I}() where {T,I} = new{T,I}(I[], I[], T[])
+end
+
+function update_accumulator!(acc::MatrixBuilderAccumulator, mb::MatrixBuilder, mapping)
+    for i in eachindex(mb.is)
+        if haskey(mapping, mb.is[i]) # skip keys that were filtered
+            push!(acc.is, mapping[mb.is[i]])
+            push!(acc.js, mb.js[i])
+            push!(acc.vs, mb.vs[i])
+        end
+    end
+end
+
+function finalize_accumulator!(
+    acc::MatrixBuilderAccumulator{T,I}, basis, sort; kwargs...
+) where {T,I}
+    n = length(basis)
+
+    # see docstring of `sparse!` for an explanation for what this is
+    klasttouch = Vector{I}(undef, n)
+    csrrowptr = Vector{I}(undef, n + 1)
+    csrcolval = Vector{I}(undef, length(acc.is))
+    csrnzval = Vector{T}(undef, length(acc.is))
+
+    matrix = SparseArrays.sparse!(
+        acc.is, acc.js, acc.vs, n, n, +,
+        klasttouch, csrrowptr, csrcolval, csrnzval,
+        acc.is, acc.js, acc.vs,
+    )
+    if sort
+        # `csrcolval` is no longer needed, so we can reuse it
+        perm = resize!(csrcolval, length(basis))
+        sortperm!(perm, basis; kwargs...)
+        permute!(matrix, perm, perm), permute!(basis, perm)
+    else
+        return matrix, basis
+    end
+end
+
+"""
+    basis_bfs(::Type{Builder}, operator, starting_basis)
+
+Internal function that performs breadth-first search (BFS) on an operator.
```

Comment:
```suggestion
    basis_bfs(::Type{Builder}, operator, starting_basis; kw_args...)

Internal function that performs breadth-first search (BFS) on an operator.

`Builder` is either [`MatrixBuilder`](@ref) or [`BasisBuilder`](@ref), which triggers building a
matrix and basis, or only the basis of addresses, respectively. 
```
I feel that the long list of keyword arguments doesn't need to be documented here, as this is an internal functions and interested people can just read the code. However, these keyword arguments are passed through from the user facing `build_basis` and `BasisSetRepresentation` and should definitely be documented there. Please check and  update these docstrings accordingly.